README.md
# Codify

A discord bot for the Code Community Discord Server.

# About

This bot was created for my comunity on discord because I wanted to provide them a fun bot with a lot of functionalities designed especially for my discord server.

# Add my bot to your discord server

[Just click this link](https://discordapp.com/api/oauth2/authorize?client_id=650637687451156481&permissions=8&scope=bot)

## Commands

Use `cc!help` for a complete list of commands. There are so many that I lost track.

## How to install

Clone the repo, run `npm install`, set a `.env` file for the following variables: `BOT_TOKEN, COMMUNITY, DATABASE_URL`, make sure that you at least set the `BOT_TOKEN` and `COMMUNITY` variables before continuing, next you can run `npm run dev` to start the bot locally.

If everything is set up correctly you should see your bot online. If you don't know how to make a discord app with a discord bot with a token make sure to follow [this guide](https://www.digitaltrends.com/gaming/how-to-make-a-discord-bot/) following step 2, 3 and 4.

## How to contribute.

It's easy, just fork this repository, add your changes to your fork and then make a pull request.

# Huge thanks to:

Every contributor, and also the members of my discord server that help me with feedback, give me ideas, report bugs, and use the bot everyday.

# Commands 

```md
[1]: Set your description - cc!description/desc/about [text]
- Set your description. Can be seen in cc!profile. Can't be longer than 1000 characters,
[2]: Shows a profile - cc!profile <user>
- Show the profile of a user. The default is the profile of the one who sends the command.,
[3]: Say something - cc!say [text]
- Make the bot say something,
[4]: Daily money - cc!daily/claim/dailyclaim/free
- You get a random amount of coins from 1 to 100, can be used once a day,
[5]: Roll the dice - cc!gamble/dive/slots [amount]
- Gamble a random amount of coins. can be between 1 and 10000,
[6]: Java is to js - cc!javatojs/javaistojs
- Sends an image that explains what java is to javascript,
[7]: Add a keyword for assassinations - cc!addkeyword [keyword] | cc!addassasinkeyword [keyword]
- Sends a request for adding a new keyword for assassinations. Costs 10k, if it gets rejected you get 5k back, and if it doesn't get any response in 24h then you get your 10k back,
[8]: Meme - cc!meme <subreddit>
- Sends a random meme from a subreddit. The default is set to r/ProgrammerHumor. Sometimes it doesn't work because the meme it got was from either an external source or a video/gif,
[9]: Top levels - cc!top/toplevel/topl
- Shows a top with user levels in the server
[10]: Create a webhook - cc!webhook/createwebhook
- Creates a webhook in that channel if the user has the manage webhooks permission or if the user has a role named `webhooks`. Then it sends a dm with the webhook link to the user,
[11]: Toggle the leveling messages in the server. - cc!toggleServerLevelUps | cc!toggleLevels
- Toggle the leveling up in the server, users will still level up, just without the message that shows the level.,
[12]: Create a poll - cc!poll [title]|[option1]|[option2]|...
- Creates a poll,
[13]: Age - cc!age
- Displays the age of the server or of your account.,
[14]: Trivia - cc!trivia
- Sends a trivia question that you can answer, if you are right you can get coins,
[15]: Till Xmas - cc!xmas
- Days left until Christmas,
[16]: Start the assasination of someone - cc!assassinate [serverid] [userid]
- You will be given a word and you will have to make your target say that word, without trying to use any kind of tricks, because there is a catch, if you manage to make that person say that word you were given they will lose 5% of their money, 3% of them will go to you and 2% will be lost, but if you cant manage to make that person say that word you will lose 5% of your money, 3% going to the target and 2% being lost.,
[17]: Disable the level up messages - cc!disablelevelupmessages/dlum
- If you are annoyed by the level up messages you can use this command and if you level up you will no longer get pinged by codify when you level up. You will still level up though,
[18]: Toggle the leveling up quotes in the server. - cc!toggleServerQuotes | cc!toggleQuotes
- Toggle the leveling up quotes in the server, the level up messages will still appear, just without quotes.
[19]: Get a token - cc!gettoken
- Sends a dm with your token for the api. More info into the dm. But the api doesnt work anymore so this command has no use for now.,
[20]: Google it - cc!googleit
- Sends an image from codebullet with just google it,
[21]: Add a quote - cc!addquote [text]
- Sends a request for adding a new quote for level up messages. Costs 10k, if it gets rejected you get 5k back, and if it doesn't get any response in 24h then you get your 10k back,
[22]: Ban a user - cc!ban [user] <reason>
- Bans a user, needs special permissions.,
[23]: Unban a user - cc!unban [userid]
- Unbans a user, needs special permissions.,
[24]: Kick a user - cc!kick [user] <reason>
- Kick a user, needs special permissions.,
[25]: Set the level up channel for your server. - cc!setlevelschannel <channelid>
- It sets the channel for were all of the level up messages will be sent by the bot. Use no argument to reset it.,
[26]: Set the welcome channel for your server. - cc!setwelcomechannel <channelid>
- It sets the channel for were all of the welcome and leave messages will be sent by the bot. Use no arguments to reset it.,
[27]: Set the audit log channel for your server. - cc!setauditchannel <channelid>
- It sets the channel for were all of the log messages will be sent by the bot. Use no arguments to reset it.
[28]: Top balance - cc!topb/topbalance/balancetop
- Shows a top with user balances of the server,
[29]: Make a joke - cc!joke
- Sends a random joke from r/jokes,
[30]: Opt out of the assassinations minigame. - cc!optoutofassassinations/oooa
- If you want to opt out of the assassinations minigame, you can use this command, nobody will be able to assassinate you, at the cost of you not being able to assassinate anyone.,
[31]: Top xp today - cc!topd/topleveltoday/topday
- Shows the top of xp gain today.,
[32]: Copypasta - cc!copypasta
- Sends a copypasta from r/copypasta,
[33]: Pay someone - cc!pay [user] [amount]
- Pay an user a certain amount of coins from your balace,
[34]: Shows the server settings - cc!serversettings
- It will show the server wide settings like the prefix or pins channel, etc.,
[35]: Anyway - cc!anyway/cc!anyways [text]
- Sends an image with Danny Devito that says "So anyway I started [text]",
[36]: Chart - cc!chart [title] | [white side] | [black side]
- Sends a chart comparing the white amount to the black amount. It's an inside server joke, the original chart showed distribution of black vs white people based on iq
[37]: Scream - cc!a
- Makes the bot scream. AAAAAAAAAAAAAAAAAAAAAAAAAAAA,
[38]: Add item to the store - cc!addstoreitem [roleid] [price] [interval in days]
- People with `Manage roles` permission can use this command to add an item to the store, aka a role that can be bought by other users,
[39]: Remove store item - cc!removestoreitem [itemid]
- Remove an item from the store, you need `Manage roles` permission for this,
[40]: Servertime - cc!servertime
- Shows the bot's time, aka the server time.,
[41]: Set the prefix for the commands. - cc!setprefix [prefix]
- It sets the prefix for the commands in the server.,
[42]: Store - cc!store
- Shows the items in the store,
[43]: Top balance globally - cc!topbg/topbalanceglobal/balancetopglobal
- Shows a top with user balances globally.,
[44]: Buy an item - cc!buy [itemid]
- Buy an item from the store of the server. You have to use the id listed in cc!store
[46]: Top levels globally - cc!topg/toplevelglobal/toplg
- Shows a top with user levels globally.,
[47]: Create a pin - cc!pin [messagelink] <description>
- Creates a pin in a special channel created by the server administrator, you need the Manage messages permission to use this command,
[48]: Top xp today globally - cc!topdg/topleveltodayglobal/topdayglobal
- Shows the top of xp gain today globally.,
[49]: Set the pins channel for your server. - cc!setpinschannel <channelid>
- It sets the channel for were the cc!pin messages will go. Use no arguments to reset it.,
[50]: Purchases - cc!purchases
- Shows what purchases you have in the server,
[51]: Unsubscribe - cc!unsubscribe [itemid]
- Unsubscribe from a role that you bought,
[52]: help - cc!help
- Shows this output,
[53]: Reactions - passive
- The bot will react to random messages, it's passive
```



CHANGELOG.md
# [1.25.0](https://github.com/Vyctor661/codify/compare/v1.24.1...v1.25.0) (2022-02-18)


### Features

* changed the formula for daily rewards ([8f54dec](https://github.com/Vyctor661/codify/commit/8f54dec6c28c571a2b2342d80b49d665024a4dd1))

## [1.24.1](https://github.com/Vyctor661/codify/compare/v1.24.0...v1.24.1) (2022-02-17)


### Bug Fixes

* another error that I spotted in heroku logs ([7cbcacd](https://github.com/Vyctor661/codify/commit/7cbcacd770a175aa2ee7601e3a2c76917bae62f2))
* broken highlighting fixed ([c043172](https://github.com/Vyctor661/codify/commit/c043172f4c4c22e0b5e0d559aeb30c3faeef20f6))
* hopefully this will fix the thing that caused the crash earlier today ([ff9ab1b](https://github.com/Vyctor661/codify/commit/ff9ab1ba37b804b1eaa4c9a9a329791291af08a7))
* message update creating update loop in audit channel ([e514438](https://github.com/Vyctor661/codify/commit/e5144389ef31a6d1568257603f53c0f470adb076))
* moving a channel around doesnt trigger the channelupdate audit a million times anymore ([e8e12fe](https://github.com/Vyctor661/codify/commit/e8e12fe5aed86be30b862390f24e132859084d42))

# [1.24.0](https://github.com/Vyctor661/codify/compare/v1.23.0...v1.24.0) (2022-02-14)


### Features

* safety for not being able to ban or kick the server owner (just in case) ([f53925e](https://github.com/Vyctor661/codify/commit/f53925e12a48a794551ad979f4ce66bcbe5b3398))

# [1.23.0](https://github.com/Vyctor661/codify/compare/v1.22.0...v1.23.0) (2022-02-14)


### Bug Fixes

* annoying error caused by updating node fetch ([4323a7f](https://github.com/Vyctor661/codify/commit/4323a7fb61efe77e6ef724a484c4035af72df508))
* bad file pathnames ([c866427](https://github.com/Vyctor661/codify/commit/c866427beed75a3cb8cab8b004f4a88081223a41))
* changed the success message because it was saying pins channel instead of welcome channel ([e18c87f](https://github.com/Vyctor661/codify/commit/e18c87fdcb4493ba88a1a3f200f0f358b7f10123))
* help being seen as an invalid command ([26611cd](https://github.com/Vyctor661/codify/commit/26611cd41e282210512b213dbe439ca24791556c))
* help command broken after full update ([cb94d7a](https://github.com/Vyctor661/codify/commit/cb94d7ae438e1f0797fa4cf48e8ad1c8e7eacdff))
* help command crashing the bot ([4452a42](https://github.com/Vyctor661/codify/commit/4452a42cc01dc8751d096ae3423233985bdf1678))
* image manipulation commands now work after update ([8bff5d3](https://github.com/Vyctor661/codify/commit/8bff5d3ae85cafba0c57e15cee5833d8849b2e13))
* invalid command now works again (I blame myself when I refactored the help command long time ago I apparently broke the invalid command thing) ([5d746c8](https://github.com/Vyctor661/codify/commit/5d746c857c5b24dab6fed50de7c2e68b18d12ae0))
* most of the errors after updating discord js, a lot is still broken and needs to be fixed ([7b97b07](https://github.com/Vyctor661/codify/commit/7b97b07168ea7fd1debf14d21e91bb648e5c2f29))
* potential bugs created by synchronicity ([1d7f99e](https://github.com/Vyctor661/codify/commit/1d7f99e13505d46f3d9e1ed40bd649db4e339822))
* tops not working after updating ([2fec7fe](https://github.com/Vyctor661/codify/commit/2fec7fe9ae4dd6d3a54a37ee300ef4205f66625c))
* trivia command crashing the bot ([cf6f046](https://github.com/Vyctor661/codify/commit/cf6f0463fcedc5e798e507ad7ee6efe3c52bfa80))
* updated the set welome channel id command suffix ([d7a0457](https://github.com/Vyctor661/codify/commit/d7a04571d577e30ddcf39f79d60240d1de668dd2))
* updated typescript, fixes a lot of errors ([ab65137](https://github.com/Vyctor661/codify/commit/ab65137e65c38509df855c7bc5c878a472fb76ac))
* webhooks command now works again ([def722f](https://github.com/Vyctor661/codify/commit/def722f2a24f63d56077c7d10735c1a5e31959a2))
* welcome leave messages fix after update ([429318f](https://github.com/Vyctor661/codify/commit/429318f2742bbd79b86175926941f4b75be7ae71))


### Features

* added the logevent to the commands that I think should have it ([2a83a47](https://github.com/Vyctor661/codify/commit/2a83a473fe86e5e9689d95fb9f3b122f42dd0160))
* audit log events ([3f799f4](https://github.com/Vyctor661/codify/commit/3f799f430847eb46487f841830fb7fd4a01855da))
* audit logs no longer ping users, just mentions them ([13921b3](https://github.com/Vyctor661/codify/commit/13921b3397e2cb27ea5e21b85945533588c9a371))
* ban command ([feecfb3](https://github.com/Vyctor661/codify/commit/feecfb31b0e9fe404b640f6588b50ecff93b9ce2))
* better looking embed ([07b5095](https://github.com/Vyctor661/codify/commit/07b509507891548b52b969ec42df1e3f96030ab2))
* ccdaily amount of money now scales with level ([d5f4819](https://github.com/Vyctor661/codify/commit/d5f4819a3b1d99cfbb6539b216af0ec187d2f01c))
* db implementation for only quotes and commands prefix ([1ceb23a](https://github.com/Vyctor661/codify/commit/1ceb23ac86d8ad345fd0badd27f592f575002ada))
* help command shows the proper prefix for the server ([e8ad041](https://github.com/Vyctor661/codify/commit/e8ad0417eb995155d0754a0f528893f5547f1222))
* implemented level up messages channl functionality ([a1449b9](https://github.com/Vyctor661/codify/commit/a1449b9825f09535950c0b41035292a5f7d47d08))
* implemented logging function ([06fcea3](https://github.com/Vyctor661/codify/commit/06fcea3a863a15d18bede8f9cb04f1a413ff0579))
* implemented no level up messages server wide ([025d310](https://github.com/Vyctor661/codify/commit/025d3101878d8b9efc1b11802249daead527abd2))
* kick command ([83e0233](https://github.com/Vyctor661/codify/commit/83e02336129fb9b1842371016e1ebe6864a6a452))
* now no args resets the audit channel ([90944f7](https://github.com/Vyctor661/codify/commit/90944f7c330face06ade58e991efa3a8bb4ba86a))
* now no args resets the pins channel ([d24e4f0](https://github.com/Vyctor661/codify/commit/d24e4f0917c60691134413d19a3c96b123d29e15))
* now no args resets the welcome channel ([347a2a4](https://github.com/Vyctor661/codify/commit/347a2a4f9e5b1bc45d163b380430b95aba0fa483))

src/modules/constants.ts
import express from "express";

export const prefix = process.env.NODE_ENVI === "development" ? "id!" : "cc!";
export const port = process.env.PORT || 5000;

export const app = express();

export const resolveArrayToOne = <T>(arr: T[] | T) => {
    if (Array.isArray(arr)) return arr[0];
    return arr;
};

export const alphabet = "abcdefghijklmnopqrstuvwxyz";
export const emojiLetters = [
    "🇦",
    "🇧",
    "🇨",
    "🇩",
    "🇪",
    "🇫",
    "🇬",
    "🇭",
    "🇮",
    "🇯",
    "🇰",
    "🇱",
    "🇲",
    "🇳",
    "🇴",
    "🇵",
    "🇶",
    "🇷",
    "🇸",
    "🇹",
    "🇺",
    "🇻",
    "🇼",
    "🇽",
    "🇾",
    "🇿"
];

export const nth = (d: number) => {
    if (d > 3 && d < 21) return "th";
    switch (d % 10) {
        case 1:
            return "st";
        case 2:
            return "nd";
        case 3:
            return "rd";
        default:
            return "th";
    }
};

export const xpMultiplier = 1;
export const jackpotMultiplier = 2;
export const maxBetLimit = 10000;
export const topXpWinXpGain = 500;
export const topXpWinMoneyGain = 2500;

export const emojis = [
    "😀",
    "😃",
    "😄",
    "😁",
    "😆",
    "😅",
    "😂",
    "🤣",
    "😊",
    "😇",
    "🙂",
    "🙃",
    "😉",
    "😌",
    "😍",
    "🥰",
    "😘",
    "😗",
    "😙",
    "😚",
    "😋",
    "😛",
    "😝",
    "😜",
    "🤪",
    "🤨",
    "🧐",
    "🤓",
    "😎",
    "🤩",
    "🥳",
    "😏",
    "😒",
    "😞",
    "😔",
    "😟",
    "😕",
    "🙁",
    "☹️",
    "😣",
    "😖",
    "😫",
    "😩",
    "🥺",
    "😢",
    "😭",
    "😤",
    "😠",
    "😡",
    "🤬",
    "🤯",
    "😳",
    "🥵",
    "🥶",
    "😱",
    "😨",
    "😰",
    "😥",
    "😓",
    "🤗",
    "🤔",
    "🤭",
    "🤫",
    "🤥",
    "😶",
    "😐",
    "😑",
    "😬",
    "🙄",
    "😯",
    "😦",
    "😧",
    "😮",
    "😲",
    "🥱",
    "😴",
    "🤤",
    "😪",
    "😵",
    "🤐",
    "🥴",
    "🤢",
    "🤮",
    "🤧",
    "😷",
    "🤒",
    "🤕",
    "🤑",
    "🤠",
    "😈",
    "👿",
    "👹",
    "👺",
    "🤡",
    "💩",
    "👻",
    "💀",
    "☠️",
    "👽",
    "👾",
    "🤖",
    "🎃",
    "😺",
    "😸",
    "😹",
    "😻",
    "😼",
    "😽",
    "🙀",
    "😿",
    "😾",
    "👋",
    "🤚",
    "🖐",
    "✋",
    "🖖",
    "👌",
    "🤏",
    "✌️",
    "🤞",
    "🤟",
    "🤘",
    "🤙",
    "👈",
    "👉",
    "👆",
    "🖕",
    "👇",
    "☝️",
    "👍",
    "👎",
    "✊",
    "👊",
    "🤛",
    "🤜",
    "👏",
    "🙌",
    "👐",
    "🤲",
    "🤝",
    "🙏",
    "✍️",
    "💅",
    "🤳",
    "💪",
    "🦾",
    "🦵",
    "🦿",
    "🦶",
    "👣",
    "👂",
    "🦻",
    "👃",
    "🧠",
    "🦷",
    "🦴",
    "👀",
    "👅",
    "👄",
    "💋",
    "🩸"
];

export const activity = (): string => {
    switch (new Date().getMonth()) {
        case 0: // Jan
            if (new Date().getDate() === 1) return "🎆 Drinking fireworks.";
            break;
        case 1: // Feb
            break;
        case 2: // Mar
            break;
        case 3: // Apr
            if (new Date().getDate() === 1) return "🤡 Drinking jokes.";
            break;
        case 4: // May
            break;
        case 5: // Jun
            return "🌈 Drinking pride.";
        case 6: // Jul
            return "🏖️ Drinking sand.";
        case 7: // Aug
            break;
        case 8: // Sept
            if (new Date().getDate() === 2) return "🎂 Drinking CC cake.";
            if (new Date().getDate() === 13) return "💻 Drinking code.";
            break;
        case 9: // Oct
            if (new Date().getDate() === 31) return "🩸 Drinking blood.";
            return "🎃 Drinking lava.";
        case 10: // Nov
            return "🥜 Drinking peanut butter.";
        case 11: // Dec
            if (new Date().getDate() === 1) return "🇷🇴 Drinking țuică.";
            if (new Date().getDate() === 25) return "🎁 Drinking gifts.";
            if (new Date().getDate() === 31) return "🍷 Drinking wine.";
            return "❄️ Drinking snow.";
    }
    return "💧 Drinking water.";
};

export const ymwdhms = (diff: number): string => {
    const yearLen = 1000 * 60 * 60 * 24 * 365;
    const monthLen = 1000 * 60 * 60 * 24 * 30;
    const weekLen = 1000 * 60 * 60 * 24 * 7;
    const dayLen = 1000 * 60 * 60 * 24;
    const hourLen = 1000 * 60 * 60;
    const minLen = 1000 * 60;

    const years = Math.floor(diff / yearLen);
    diff %= yearLen;
    const months = Math.floor(diff / monthLen);
    diff %= monthLen;
    const weeks = Math.floor(diff / weekLen);
    diff %= weekLen;
    const days = Math.floor(diff / dayLen);
    diff %= dayLen;
    const hours = Math.floor(diff / hourLen);
    diff %= hourLen;
    const minutes = Math.floor(diff / minLen);
    diff %= minLen;
    const seconds = Math.floor(diff / 1000);

    return `${years} Year${years > 1 ? "s" : ""} ${months} Month${
        months > 1 ? "s" : ""
    } ${weeks} Week${weeks > 1 ? "s" : ""} ${days} Day${
        days > 1 ? "s" : ""
    } ${hours} Hour${hours > 1 ? "s" : ""} ${minutes} Minute${
        minutes > 1 ? "s" : ""
    } and ${seconds} Second${seconds > 1 ? "s" : ""}`;
};

export const alphanumericRegx = /[^A-Za-z0-9]/g;
